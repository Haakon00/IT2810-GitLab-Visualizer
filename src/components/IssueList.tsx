import { useContext, useEffect, useState } from "react";
import Issue from "./Issue/Issue";
import getIssues from "../functions/getIssues";
import { DarkModeContext } from "./DarkModeProvider";
export type issueInfo = {
  title: string;
  state: string;
  iid: number | string;
  assignee: {
    avatar_url: string;
    state: string;
    name: string;
  };
};

//Returns all the issues with status
const IssueList = () => {
  const [issues, setIssues] = useState<issueInfo[]>([]);

  const [all, setAll] = useState<Boolean>(true);
  const [opened, setOpened] = useState<Boolean>(false);
  const [closed, setClosed] = useState<Boolean>(false);

  useEffect(() => {
    getIssues().then(async (response) => {
      if (response.status === 200) {
        setIssues((await response.json()) as issueInfo[]);
      }
    });
  }, []);

  const openedIssues = issues.filter((issues) => issues.state === "opened");
  const closedIssues = issues.filter((issues) => issues.state === "closed");

  const darkMode = useContext(DarkModeContext);

  return (
    <div>
      <div
        className={
          darkMode.darkMode
            ? "text-3xl font-bold font-[verdana] text-white"
            : "text-3xl font-bold font-[verdana] text-black"
        }
      >
        <h2 className="p-2 h-[5vh] lg:pt-3 sm:text-center">Issues</h2>
        <div className="h-[5vh] text-lg flex justify-evenly">
          <div className="flex items-center">
            <button onClick={() => document.getElementById("all")!.click()}>
              All&nbsp;
            </button>
            <input
              type="radio"
              name="radio-1"
              className="radio radio-accent"
              id="all"
              onClick={() => {
                setAll(true);
                setOpened(false);
                setClosed(false);
              }}
              defaultChecked
            />
          </div>
          <div className="flex items-center">
            <button onClick={() => document.getElementById("open")!.click()}>
              Open&nbsp;
            </button>
            <input
              type="radio"
              name="radio-1"
              className="radio radio-accent"
              id="open"
              onClick={() => {
                setAll(false);
                setOpened(true);
                setClosed(false);
              }}
            />
          </div>
          <div className="flex items-center">
            <button onClick={() => document.getElementById("closed")!.click()}>
              Closed&nbsp;
            </button>
            <input
              type="radio"
              name="radio-1"
              className="radio radio-accent"
              id="closed"
              onClick={() => {
                setAll(false);
                setOpened(false);
                setClosed(true);
              }}
            />
          </div>
        </div>
        <div
          className={"lg:h-[80vh] lg:overflow-y-scroll overflow-x-hidden"}
          id="style-1"
        >
          {issues.length === 0 ? (
            <h1>Invalid API-key or token</h1>
          ) : (
            <div>
              {opened
                ? openedIssues.map((issue) => {
                    return <Issue issueInfo={issue} key={issue.iid}></Issue>;
                  })
                : closed
                ? closedIssues.map((issue) => {
                    return <Issue issueInfo={issue} key={issue.iid}></Issue>;
                  })
                : issues.map((issue) => {
                    return <Issue issueInfo={issue} key={issue.iid}></Issue>;
                  })}
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

export default IssueList;
