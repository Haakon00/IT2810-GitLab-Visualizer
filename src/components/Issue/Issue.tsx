import { useContext } from "react";
import { DarkModeContext } from "../DarkModeProvider";
import { issueInfo } from "../IssueList";

function Issue(props: { issueInfo: issueInfo }) {
	const darkMode = useContext(DarkModeContext);
	return (
		<div
			className={
				darkMode.darkMode
					? "card w-96 bg-slate-800 shadow-xl text-center text-white my-3 text-base"
					: "card w-96 bg-[#FEF08A] shadow-xl text-center text-black my-3 text-base"
			}
		>
			<div data-testid="title" className="mt-1 mx-2">
				{props.issueInfo.title}
			</div>
			<div data-testid="status" className="font-normal">
				Status: {props.issueInfo.state}
			</div>
			{props.issueInfo.assignee && (
				<div data-testid="responsible" className="flex flex-col font-normal items-center">
					Ansvarlig: {props.issueInfo.assignee.name}
					<img
						className="w-1/12 object-cover my-4"
						src={props.issueInfo.assignee.avatar_url}
						alt="assignee avatar"
					/>
				</div>
			)}
		</div>
	);
}

export default Issue;
