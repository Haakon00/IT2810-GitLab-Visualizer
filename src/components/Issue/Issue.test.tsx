import { screen, render } from "@testing-library/react";
import renderer from "react-test-renderer";
import Issue from "./Issue";
import { issueInfo } from "../IssueList";

beforeAll(() => {});

test("Renders Issue props", () => {
	const issueInfo: issueInfo = {
		title: "tittel",
		state: "closed",
		iid: -1,
		assignee: { avatar_url: "test.com", state: "closed", name: "Alfonzo" },
	};

	render(<Issue issueInfo={issueInfo} />);
	const responsible = screen.getByTestId("responsible");
	expect(responsible.textContent).toBe("Ansvarlig: Alfonzo");
	const status = screen.getByTestId("status");
	expect(status.textContent).toBe("Status: closed");
	const title = screen.getByTestId("title");
	expect(title.textContent).toBe("tittel");
});

test("Snapshot test of Issue card", () => {
	const issueInfo: issueInfo = {
		title: "tittel",
		state: "closed",
		iid: -1,
		assignee: { avatar_url: "test.com", state: "closed", name: "Alfonzo" },
	};
	const tree = renderer.create(<Issue issueInfo={issueInfo} />).toJSON();
	expect(tree).toMatchSnapshot();
});
