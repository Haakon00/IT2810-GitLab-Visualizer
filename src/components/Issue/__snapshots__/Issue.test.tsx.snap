// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test of Issue card 1`] = `
<div
  className="card w-96 bg-[#FEF08A] shadow-xl text-center text-black my-3 text-base"
>
  <div
    className="mt-1 mx-2"
    data-testid="title"
  >
    tittel
  </div>
  <div
    className="font-normal"
    data-testid="status"
  >
    Status: 
    closed
  </div>
  <div
    className="flex flex-col font-normal items-center"
    data-testid="responsible"
  >
    Ansvarlig: 
    Alfonzo
    <img
      alt="assignee avatar"
      className="w-1/12 object-cover my-4"
      src="test.com"
    />
  </div>
</div>
`;
